File: packages/pluto/src/kvstore.ts
import {
  IResource,
  IResourceCapturedProps,
  IResourceClientApi,
  IResourceInfraApi,
  PlatformType,
  simulator,
  utils,
} from "@plutolang/base";
import { aws, k8s } from "./clients";

/**
 * The options for instantiating an infrastructure implementation class or a client implementation
 * class.
 */
export interface KVStoreOptions {}

export interface IKVStoreRegularApi {
  readonly awsTableName?: string;
  readonly awsPartitionKey?: string;
}

/**
 * Define the access methods for KVStore that operate during runtime.
 */
export interface IKVStoreClientApi extends IResourceClientApi {
  get(key: string): Promise<string>;
  set(key: string, val: string): Promise<void>;
}

/**
 * Define the methods for KVStore, which operate during compilation.
 */
export interface IKVStoreInfraApi extends IResourceInfraApi {}

export interface IKVStoreCapturedProps extends IResourceCapturedProps {}

/**
 * Construct a type that includes all the necessary methods required to be implemented within the
 * client implementation class of a resource type.
 */
export type IKVStoreClient = IKVStoreClientApi & IKVStoreCapturedProps & IKVStoreRegularApi;

/**
 * Construct a type that includes all the necessary methods required to be implemented within the
 * infrastructure implementation class of a resource type.
 */
export type IKVStoreInfra = IKVStoreInfraApi & IKVStoreCapturedProps;

// TODO: abstract class
export class KVStore implements IResource {
  constructor(name: string, opts?: KVStoreOptions) {
    name;
    opts;
    throw new Error(
      "cannot instantiate this class, instead of its subclass depending on the target runtime."
    );
  }

  public static buildClient(name: string, opts?: KVStoreOptions): IKVStoreClient {
    const platformType = utils.currentPlatformType();
    switch (platformType) {
      case PlatformType.AWS:
        return new aws.DynamoKVStore(name, opts);
      case PlatformType.K8s:
        return new k8s.RedisKVStore(name, opts);
      case PlatformType.Simulator: {
        if (!process.env.PLUTO_SIMULATOR_URL) throw new Error("PLUTO_SIMULATOR_URL doesn't exist");
        const resourceId = utils.genResourceId(KVStore.fqn, name);
        return simulator.makeSimulatorClient(process.env.PLUTO_SIMULATOR_URL!, resourceId);
      }
      default:
        throw new Error(`not support this runtime '${platformType}'`);
    }
  }

  public static fqn = "@plutolang/pluto.KVStore";
}

export interface KVStore extends IResource, IKVStoreClient, IKVStoreInfra {}

File: packages/pluto-py/pluto_client/kvstore.py
from dataclasses import dataclass
from typing import Optional
from pluto_base.resource import (
    IResource,
    IResourceCapturedProps,
    IResourceClientApi,
    IResourceInfraApi,
)
from pluto_base.platform import PlatformType
from pluto_base import utils
from .utils import create_simulator_client


@dataclass
class KVStoreOptions:
    pass


class IKVStoreRegularApi:
    @property
    def aws_table_name(self) -> str:
        raise NotImplementedError

    @property
    def aws_partition_key(self) -> str:
        raise NotImplementedError


class IKVStoreClientApi(IResourceClientApi):
    def get(self, key: str) -> str:
        raise NotImplementedError

    def set(self, key: str, val: str) -> None:
        raise NotImplementedError


class IKVStoreInfraApi(IResourceInfraApi):
    pass


class IKVStoreCapturedProps(IResourceCapturedProps):
    pass


class IKVStoreClient(IKVStoreClientApi, IKVStoreCapturedProps, IKVStoreRegularApi):
    pass


class IKVStoreInfra(IKVStoreInfraApi, IKVStoreCapturedProps):
    pass


class KVStore(IResource, IKVStoreClient, IKVStoreInfra):
    fqn = "@plutolang/pluto.KVStore"

    def __init__(self, name: str, opts: Optional[KVStoreOptions] = None):
        platform_type = utils.current_platform_type()
        if platform_type == PlatformType.AWS:
            from .clients import aws

            self._client = aws.DynamoKVStore(name, opts)

        elif platform_type == PlatformType.Simulator:
            resource_id = utils.gen_resource_id(KVStore.fqn, name)
            self._client = create_simulator_client(resource_id)  # type: ignore

        else:
            raise ValueError(f"not support this runtime '{platform_type}'")
